### DISCUSSION TITLE
How to update an entity by primary key

### DISCUSSION THREAD

**Discussion by StormstoutY** [milvus-io/milvus#26828] [Q&A and General discussion]
I am using milvus v2.3.0. Now I need to display all entities of a collection and provide corresponding delete and update functions.

If I need to update an entity,  can I only delete this entity through the primary key, and then insert new data with the same primary key? Since autoID is set to True for primary key fields in my collection, I cannot use upsert() method.

For delete function, I noticed that in the document of delete entities, it is mentioned that "Frequent deletion operations will impact the system performance". What adverse effects will frequent deletion operations have? Is there any way to solve this effect?

**Comment by yhmo** [comment]
For auto-id collection, after you call insert() api, it returns the ids generated by server. You can get the auto-id by its response:
```
resp = collection.insert(data)
ids = mr.primary_keys
print(ids)
```
With the returned id, you can call upsert or delete by primary key.

If you didn't store the auto-id in some place, you can call query() to return ids with special conditions, and use the ids to call delete() or upsert().

Frequent deletion will lead to these problems:
1. Memory/disk usage increases. In milvus, delete is soft-delete, the deleted items are not cleaned from memory/disk instantly, they are still occupying the space, waiting for gc process. Once GC process is triggered, the data nodes will check all the segments. If a segment has been deleted more than 10% items, the data node will compact the segment. The compaction process is: extract the non-delete items to be a new segment, and mark the old segment as soft-delete, the old segment will be permanently deleted in one hour. You will observe disk usage increase in GC process.
2. Slightly slow down the search performance. Since soft-delete items are still in memory before the segment is compacted, all the search action will have to handle the soft-delete items, exclude them from search scope.

This is internal designed machinery, all the functions have cost.

**Reply by StormstoutY** [reply]
Thank you so much, I call query() to get all the entities of the collection, then I use upsert by primary key, but an exception occurs like "pymilvus.exceptions.UpsertAutoIDTrueException: <UpsertAutoIDTrueException: (code=1, message=Upsert don't support autoid == true)>". It seems that auto-id collection can only update an entity by delete and insert. And since auto-id may use something like timestamp as the primary key, I can't use the original primary key to insert.

**Comment by xiaofan-luan** [comment]
> I am using milvus v2.3.0. Now I need to display all entities of a collection and provide corresponding delete and update functions.
> 
> If I need to update an entity, can I only delete this entity through the primary key, and then insert new data with the same primary key? Since autoID is set to True for primary key fields in my collection, I cannot use upsert() method.
> 
> For delete function, I noticed that in the document of delete entities, it is mentioned that "Frequent deletion operations will impact the system performance". What adverse effects will frequent deletion operations have? Is there any way to solve this effect?

You can't do anything without know it's primary key. 

1. you can query by expr and find the primary key, but it's better to store it somewhere.
2. with PK, you can do upsert or delete

**Reply by StormstoutY** [reply]
Thank you, I try to call upsert() using the primary key, but it seems that upsert() is not applicable when autoID is set to True for primary key fields just as the milvus docs says

**Comment by StormstoutY** [comment]
I wonder how auto-id is calculated and  do I need to call flush() or create index after every deletion and insertion operation?

**Reply by yhmo** [reply]
Auto-id is generated by the current timestamp, but it is not a simple timestamp, it contains a value of a system counter.
We don't recommend users to call flush() manually, just let Milvus to automatically trigger flush actions. Sometimes we call flush() when we know the data has been completely inserted for a collection, to make sure the data is persisted into storage.

**Reply by StormstoutY** [reply]
I get it. Thank u!!!
